src/server/bitops_family.cc

```diff
@@ -0,0 +1,304 @@
+void BitFieldRo(CmdArgList args, ConnectionContext* cntx);
+void BitOp(CmdArgList args, ConnectionContext* cntx);
+void GetBit(CmdArgList args, ConnectionContext* cntx);
+void SetBit(CmdArgList args, ConnectionContext* cntx);
+
+OpResult<std::string> ReadValue(const OpArgs& op_args, std::string_view key);
+OpResult<bool> ReadValueBitsetAt(const OpArgs& op_args, std::string_view key, uint32_t offset);
+std::string GetString(EngineShard* shard, const PrimeValue& pv);
+bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry);
+
+// ------------------------------------------------------------------------- //
+//  Bits manipulation functions
+constexpr int32_t GetBitIndex(uint32_t offset) noexcept {
+  return offset % OFFSET_FACTOR;
+}
+
+constexpr int32_t GetNormalizedBitIndex(uint32_t offset) noexcept {
+  return (OFFSET_FACTOR - 1) - GetBitIndex(offset);
+}
+
+constexpr int32_t GetByteIndex(uint32_t offset) noexcept {
+  return offset / OFFSET_FACTOR;
+}
+
+uint8_t GetByteValue(const std::string& str, uint32_t offset) {
+  return static_cast<uint8_t>(str[GetByteIndex(offset)]);
+}
+
+constexpr bool CheckBitStatus(uint8_t byte, uint32_t offset) {
+  return byte & (0x1 << offset);
+}
+
+// return true if bit is on
+bool GetBitValue(const std::string& entry, uint32_t offset) {
+  const auto byte_val{GetByteValue(entry, offset)};
+  const auto index{GetNormalizedBitIndex(offset)};
+  return CheckBitStatus(byte_val, index);
+}
+
+bool GetBitValueSafe(const std::string& entry, uint32_t offset) {
+  return ((entry.size() * OFFSET_FACTOR) > (std::size_t)offset) ? GetBitValue(entry, offset)
+                                                                : false;
+}
+
+constexpr uint8_t TurnBitOn(uint8_t on, uint32_t offset) {
+  return on |= 1 << offset;
+}
+
+constexpr uint8_t TunBitOff(uint8_t on, uint32_t offset) {
+  return on &= ~(1 << offset);
+}
+
+bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry) {
+  // we need to return the old value after setting the value for offset
+  const auto old_value{GetBitValue(*entry, offset)};  // save this as the return value
+  auto byte{GetByteValue(*entry, offset)};
+  std::bitset<8> bits{byte};
+  const auto bit_index{GetNormalizedBitIndex(offset)};
+  byte = bit_value ? TurnBitOn(byte, bit_index) : TunBitOff(byte, bit_index);
+  (*entry)[GetByteIndex(offset)] = byte;
+  return old_value;
+}
+
+// ------------------------------------------------------------------------- //
+// Helper functions to access the data or change it
+
+class OverrideValue {
+  EngineShard* shard_ = nullptr;
+  DbIndex index_ = 0;
+
+ public:
+  explicit OverrideValue(const OpArgs& args) : shard_{args.shard}, index_{args.db_ind} {
+  }
+
+  OpResult<bool> Set(std::string_view key, uint32_t offset, bool bit_value);
+};
+
+OpResult<bool> OverrideValue::Set(std::string_view key, uint32_t offset, bool bit_value) {
+  auto& db_slice = shard_->db_slice();
+  auto index = index_;
```

Comment:
here is a good example, where I think `auto` does not help readability.
`index` is a too generic word, so either call it `db_index` or use `DbIndex` type to denote that it's a db index.

Reply:
ok